### PR TITLE
buffered I/O

### PR DESCRIPTION
**Pull request by ydnar**
This branch streamlines the XML encoding and decoding for EPP by avoiding the use of `xml.Marshal` and `Unmarshal` where possible, using pre-generated XML byte slices and reusing buffers.

### CODE REVIEW

File: bool.go [hunk 1]
```diff
@@ -34,7 +45,14 @@ func (b Bool) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
 	return nil
 }
 
-var (
-	openAngle  = xml.CharData("<")
-	closeAngle = xml.CharData(">")
-)
+// MarshalXMLAttr implements the xml.MarshalerAttr interface.
+// Attributes will be serialized with a value of "1".
```
Comment (cee-dub) [inline bool.go]: or "0"

File: check.go [hunk 2]
```diff
@@ -52,3 +57,27 @@ func (c *Conn) CheckDomain(domains ...string) (*DomainCheck, error) {
...
+		}
+		switch node := t.(type) {
+		case xml.StartElement:
+			fmt.Printf("StartElement: %s %s\n", node.Name.Space, node.Name.Local)
+		case xml.EndElement:
+			fmt.Printf("EndElement: %s %s\n", node.Name.Space, node.Name.Local)
+		case xml.CharData:
+			fmt.Printf("CharData: %s\n", string(node))
+		}
+		fmt.Printf("Stack: %+v\n", d.Stack)
```
Comment (cee-dub) [inline check.go]: These `Printf`s look like debug. Intentional?

Comment (ydnar) [inline check.go]: Yes. This function is currently a spike and unused in production.

File: conn.go [hunk 3]
```diff
@@ -1,102 +1,109 @@
...
-	_, err = c.Conn.Write(p)
+	_, err = c.Conn.Write(x)
 	return err
 }
 
-// xmlHeader is a byte-slice representation of the
-// standard XML header. Declared as a global to relieve GC pressure.
-var xmlHeader = []byte(xml.Header)
-
 // readMessage reads a single EPP response from c and parses the XML into req.
```
Comment (cee-dub) [inline conn.go]: Change to "parses the XML into `msg`"?

File: decoder_test.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		Foo     string   `xml:"foo"`
+	}{}
+
+	buf.Reset()
+	buf.Write([]byte(`<hello><foo>foo</foo></hello>`))
+	d.Reset()
+	st.Expect(t, d.InputOffset(), int64(0))
+	d.Decode(&v)
+	st.Expect(t, v.Foo, "foo")
+	st.Expect(t, d.InputOffset(), int64(29))
```
Comment (cee-dub) [inline decoder_test.go]: Maybe connect 29 with the length of your test XML?

File: decoder_test.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	</response>
+</epp>`)
+
+	d := NewDecoder(bytes.NewBuffer(x))
+	var msg message
+	err := d.DecodeMessage(&msg)
+	st.Expect(t, err, nil)
+	st.Reject(t, msg.Response, nil)
+	st.Expect(t, msg.Response.ResponseData.DomainCheckData.Results[0].Domain.Domain, "good.memorial")
+	st.Expect(t, msg.Response.ResponseData.DomainCheckData.Results[0].Domain.IsAvailable, true)
```
Comment (cee-dub) [inline decoder_test.go]: This is begging for a method `msg.Results() []Result` that hides away `msg.Response.ResponseData.DomainCheckData.Results`. :snowboarder:

File: greeting.go [hunk 6]
```diff
@@ -1,87 +1,89 @@
...
+				if err != nil {
+					return err
+				}
+			}
 
-// A ErrGreetingNotFound is reported when a <greeting> message is expected but not found.
-var ErrGreetingNotFound = errors.New("missing <greeting> message")
+		case xml.EndElement:
+			// Escape early (skip remaining XML)
+			if node.Name.Local == "svcMenu" &&
```
Comment (cee-dub) [inline greeting.go]: Love it. Is the ordering reliable?

File: greeting.go [hunk 7]
```diff
@@ -1,87 +1,89 @@
...
+		case xml.CharData:
+			e := d.Element()
+			if e == nil {
+				break
+			}
+			switch e.Name.Local {
+			case "svID":
+				g.ServerName = string(node)
+			case "version":
+				g.Versions = append(g.Versions, string(node))
```
Comment (cee-dub) [inline greeting.go]: Do we use these values upstream? Thinking about why we'd need to decode them at all.

File: session.go [hunk 8]
```diff
@@ -1,39 +1,64 @@
...
+	}
+	err = c.flushDataUnit()
+	if err != nil {
+		return err
 	}
 	msg := message{}
 	return c.readMessage(&msg)
 }
+
+func encodeLogin(buf *bytes.Buffer, user, password, newPassword, version, language string, objects, extensions []string) error {
```
Comment (cee-dub) [inline session.go]: Custom encoders: always faster than reflection!

File: check.go (hunk 2 above)
Comment (cee-dub) [inline check.go]: Confirmed that running `TestDecodeCheckDomainResponse` produces lots of debugging output.

File: decoder_test.go (hunk 4 above)
Comment (ydnar) [inline decoder_test.go]: Good idea.

File: xml.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import "encoding/xml"
+
+const (
+	startEPP = `<epp xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:ietf:params:xml:ns:epp-1.0 epp-1.0.xsd" xmlns="urn:ietf:params:xml:ns:epp-1.0">`
+	endEPP   = `</epp>`
+)
+
+var (
+	xmlHeader        = []byte(xml.Header)
+	xmlPrefix        = []byte(xml.Header + startEPP)
```
Comment (cee-dub) [inline xml.go]: Value is never used.

File: decoder_test.go (hunk 5 above)
Comment (ydnar) [inline decoder_test.go]: message will ultimately go away. Replacing with a sane interface and faster parser.

File: greeting.go (hunk 6 above)
Comment (ydnar) [inline greeting.go]: Not necessarily, hence the guards.

File: xml.go [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+	xmlSuffix        = []byte(endEPP)
+	xmlCommandPrefix = []byte(xml.Header + startEPP + `<command>`)
+	xmlCommandSuffix = []byte(`</command>` + endEPP)
```
Comment (cee-dub) [inline xml.go]: Huh! This is really interesting evaluation ordering: the `+` is inside the `[]byte()` type conversion, and ``` `</command>` ``` is a raw string in that context, but it compiles and you end up with a longer `[]byte`.

File: greeting.go (hunk 7 above)
Comment (ydnar) [inline greeting.go]: ServerName useful for debugging. Objects and Extensions repeated back in Login command.

File: xml.go (hunk 10 above)
Comment (cee-dub) [inline xml.go]: And `endEPP` is already a []byte in there.

Comment (ydnar) [inline xml.go]: Trying to minimize runtime byte slice allocations.

Comment (ydnar) [inline xml.go]: endEPP is a string

Comment (cee-dub) [inline xml.go]: Derp, cont it confused with `xmlSuffix`

**Pull request merged**